File: src/main/java/com/prmncr/normativecontrol/controllers/DocumentProcessingController.java
package com.prmncr.normativecontrol.controllers;

import com.fasterxml.jackson.core.JsonProcessingException;
import com.prmncr.normativecontrol.dtos.Error;
import com.prmncr.normativecontrol.dtos.State;
import com.prmncr.normativecontrol.services.DocumentManager;
import lombok.AllArgsConstructor;
import lombok.val;
import org.springframework.core.io.ByteArrayResource;
import org.springframework.core.io.Resource;
import org.springframework.http.HttpStatus;
import org.springframework.web.bind.annotation.*;
import org.springframework.web.multipart.MultipartFile;
import org.springframework.web.server.ResponseStatusException;

import java.io.IOException;
import java.util.Collections;
import java.util.List;
import java.util.Map;

@CrossOrigin
@RestController
@RequestMapping("documents")
@AllArgsConstructor
public class DocumentProcessingController {
    private final DocumentManager documentsManager;

    @GetMapping("state")
    public Map<String, State> getState(@RequestParam(value = "id") String id) {
        val state = documentsManager.getState(id);
        if (state == null) {
            throw new ResponseStatusException(HttpStatus.NOT_FOUND, "Document not found");
        }
        return Collections.singletonMap("state", state);
    }

    @PostMapping("upload")
    @ResponseStatus(HttpStatus.ACCEPTED)
    public Map<String, String> uploadDocument(@RequestParam("file") MultipartFile file) {
        if (file == null) {
            throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "Required arguments were empty");
        }
        val filename = file.getOriginalFilename();
        if (filename == null || filename.equals("")) {
            throw new ResponseStatusException(HttpStatus.BAD_REQUEST, "Required arguments were empty");
        }
        val extension = filename.split("\\.");
        if (!extension[1].equals("docx")) {
            throw new ResponseStatusException(HttpStatus.UNPROCESSABLE_ENTITY, "Can not process this document");
        }
        byte[] bytes;
        try {
            bytes = file.getBytes();
        } catch (IOException e) {
            throw new ResponseStatusException(HttpStatus.UNPROCESSABLE_ENTITY, "Can not process this document");
        }
        return Collections.singletonMap("id", documentsManager.addToQueue(bytes));
    }

    @GetMapping("errors")
    public Map<String, List<Error>> getErrors(@RequestParam(value = "id") String id) throws JsonProcessingException {
        val data = documentsManager.getData(id);
        if (data == null) {
            throw new ResponseStatusException(HttpStatus.NOT_FOUND, "Document not found");
        }
        return Collections.singletonMap("errors", data.getDeserializedErrors());
    }

    @GetMapping("file")
    public Resource getFile(@RequestParam(value = "id") String id) {
        val file = documentsManager.getFile(id);
        if (file == null) {
            throw new ResponseStatusException(HttpStatus.NOT_FOUND, "Document not found");
        }
        return new ByteArrayResource(file.getFile());
    }

    @GetMapping("drop-database")
    @ResponseStatus(HttpStatus.OK)
    public void dropDatabase() {
        documentsManager.dropDatabase();
    }
}

File: src/main/kotlin/com/prmncr/normativecontrol/controllers/DocumentProcessingController.kt
package com.prmncr.normativecontrol.controllers

import com.fasterxml.jackson.core.JsonProcessingException
import com.prmncr.normativecontrol.services.DocumentManager
import org.springframework.http.HttpStatus
import org.springframework.http.ResponseEntity
import org.springframework.web.bind.annotation.*
import org.springframework.web.multipart.MultipartFile
import java.io.IOException

@CrossOrigin
@RestController
@RequestMapping("api/documents")
class DocumentProcessingController(private var documentsManager: DocumentManager) {
    @GetMapping("state")
    @ResponseBody
    fun getState(@RequestParam(value = "id") id: String): ResponseEntity<Any> {
        val s = documentsManager.getState(id)
        return if (s != null) ResponseEntity(object : Any() {
            val state = s
        }, HttpStatus.OK) else ResponseEntity(HttpStatus.NOT_FOUND)
    }

    @GetMapping("result")
    @ResponseBody
    fun getResult(@RequestParam(value = "id") id: String): ResponseEntity<Any> {
        val r = documentsManager.getResult(id)
        return if (r != null) ResponseEntity(object : Any() {
            val result = r
        }, HttpStatus.OK) else ResponseEntity(HttpStatus.NOT_FOUND)
    }

    @PostMapping("upload")
    @ResponseBody
    fun uploadDocument(@RequestParam("file") file: MultipartFile): ResponseEntity<Any> {
        val filename = file.originalFilename
        if (filename == null || filename == "") {
            return ResponseEntity(HttpStatus.UNPROCESSABLE_ENTITY)
        }
        val extension = filename.split("\\.".toRegex()).dropLastWhile { it.isEmpty() }.toTypedArray()
        if (extension[1] != "docx") {
            return ResponseEntity(HttpStatus.UNPROCESSABLE_ENTITY)
        }
        val bytes: ByteArray = try {
            file.bytes
        } catch (e: IOException) {
            return ResponseEntity(HttpStatus.UNPROCESSABLE_ENTITY)
        }
        val documentId = documentsManager.addToQueue(bytes)
        return ResponseEntity(object : Any() {
            val id = documentId
        }, HttpStatus.ACCEPTED)
    }

    @GetMapping("result/{id}")
    @ResponseBody
    fun loadResult(@PathVariable(value = "id") id: String): ResponseEntity<Any> {
        var file: Any? = null
        try {
            file = documentsManager.getFile(id)
        } catch (e: JsonProcessingException) {
            ResponseEntity<Any>(HttpStatus.INTERNAL_SERVER_ERROR)
        }
        return if (file != null) ResponseEntity(file, HttpStatus.OK) else ResponseEntity(HttpStatus.NOT_FOUND)
    }

    @PatchMapping("result")
    @ResponseBody
    fun saveResult(@RequestParam(value = "id") id: String): ResponseEntity<String> {
        try {
            documentsManager.saveResult(id)
        } catch (e: NullPointerException) {
            return ResponseEntity(HttpStatus.NOT_FOUND)
        } catch (e: JsonProcessingException) {
            return ResponseEntity(HttpStatus.INTERNAL_SERVER_ERROR)
        }
        return ResponseEntity(HttpStatus.OK)
    }

    @DeleteMapping("result")
    @ResponseBody
    fun deleteResult(@RequestParam(value = "id") id: String): ResponseEntity<String> {
        documentsManager.delete(id)
        return ResponseEntity(HttpStatus.OK)
    }

    @GetMapping("drop-database")
    @ResponseBody
    fun dropDatabase(): ResponseEntity<String> {
        documentsManager.dropDatabase()
        return ResponseEntity(HttpStatus.OK)
    }
}